### PR TITLE
Add run safe method that reads stderr and stdout regardless of panic

### PR DESCRIPTION
**Pull request by zcstarr**
PR here rearranges Run to properly handle a run command panic to read stdout and stderr. Previously the if a Run command panic'd it would not read stderr and stdout causing subsequent tests to read the output.

@kalman thanks for pairing/helping get this sorted!

### CODE REVIEW

**Comment by aboodman** [comment]
What was the case that it panicked during?

**Comment by zcstarr** [comment]
The panic case, occurs if a function under test panics. For instance in noms log test, catching that a command exits is done via overriding exit and triggering a panic  state to catch what happened. 
https://github.com/attic-labs/noms/blob/master/cmd/noms/noms_log_test.go#L61 is an example .

**Comment by aboodman** [comment]
I did something similar, but different for poke_test. I didn't know that the noms_log_test.go had already solved this problem.

Let's formalize this a bit more ... what if we add a util to go/util that is just a single function Exit(), which can be overridden in test. Then ClientTestUtils can override this to panic, and we can change ClientTestUtils.Run() to always work as RunSafe does here.

IOW, I don't think there's any reason to ever call Run() instead of RunSafe().

**Comment by aboodman** [comment]
Sorry, on second thought, more detailed proposal:

1. Factor out `go/util/exit`
2. Change ClientTestSuite to override `/go/util/exit::Exit()` and turn it into a panic
3. Rename current `Run()` method to `MustRun()` (this is the naming convention that means: this will panic if something goes wrong)
4. Add the new method in this PR and call it just `Run()` (since typically in Go functions that can fail return errors, so the signature here is "more normal" than the existing one)

**Comment by aboodman** [comment]
@kalman wdyt? ^

**Comment by kalman** [comment]
I agree

**Comment by zcstarr** [comment]
@aboodman  PTAL

File: samples/go/noms-merge/main_test.go [hunk 1]
```diff
@@ -65,7 +65,7 @@ func (s *testSuite) TestWin() {
 	})
 
 	var mainErr error
-	stdout, stderr := s.Run(func() { mainErr = nomsMerge() }, []string{"--quiet=true", "--parent=" + p, s.LdbDir, l, r})
+	stdout, stderr, _ := s.Run(func() { mainErr = nomsMerge() }, []string{"--quiet=true", "--parent=" + p, s.LdbDir, l, r})
```
Comment (kalman) [inline samples/go/noms-merge/main_test.go]: Either use MustRun or handle any error. Any reason why not the former?

File: samples/go/noms-merge/main_test.go [hunk 2]
```diff
@@ -110,7 +110,7 @@ func (s *testSuite) TestLose() {
 
 	for _, c := range cases {
 		var mainErr error
-		stdout, _ := s.Run(func() { mainErr = nomsMerge() }, c.args)
+		stdout, _, _ := s.Run(func() { mainErr = nomsMerge() }, c.args)
```
Comment (kalman) [inline samples/go/noms-merge/main_test.go]: Same here

File: samples/go/poke/main_test.go [hunk 3]
```diff
@@ -47,10 +47,10 @@ func (s *testSuite) TestWin() {
 	}
 
 	for k, v := range changes {
-		stdout, stderr := s.Run(main, []string{sp, k, v})
+		stdout, stderr, err := s.Run(main, []string{sp, k, v})
 		s.Equal("", stdout)
 		s.Equal("", stderr)
-		s.Equal(0, s.ExitStatus)
+		s.Equal(nil, err)
```
Comment (kalman) [inline samples/go/poke/main_test.go]: You can do `s.NoError`

File: go/util/clienttest/client_test_suite.go [hunk 4]
```diff
@@ -43,7 +48,15 @@ func (suite *ClientTestSuite) TearDownSuite() {
 	defer d.Chk.NoError(os.RemoveAll(suite.TempDir))
 }
 
-func (suite *ClientTestSuite) Run(m func(), args []string) (stdout string, stderr string) {
+func (suite *ClientTestSuite) MustRun(m func(), args []string) (stdout string, stderr string) {
```
Comment (kalman) [inline go/util/clienttest/client_test_suite.go]: Let's document this (`// MustRun runs...`)

File: go/util/clienttest/client_test_suite.go [hunk 5]
```diff
@@ -43,7 +48,15 @@ func (suite *ClientTestSuite) TearDownSuite() {
...
+	var err interface{}
+	if stdout, stderr, err = suite.Run(m, args); err != nil {
+		panic(err)
+	}
+	return
+}
+
+func (suite *ClientTestSuite) Run(m func(), args []string) (stdout string, stderr string, mainErr interface{}) {
```
Comment (kalman) [inline go/util/clienttest/client_test_suite.go]: and this

File: go/util/clienttest/client_test_suite.go [hunk 6]
```diff
@@ -23,6 +23,10 @@ type ClientTestSuite struct {
 	err        *os.File
 }
 
+type ExitError struct {
+	Code int
+}
```
Comment (kalman) [inline go/util/clienttest/client_test_suite.go]: If you define an `Error()` method then callers will be able to pass `ExitError` around as an `error` not just an `interface{}`. I.e. currently:
```go
var err interface{} = ExitError{1}
_, ok := err.(error) // ok is false
```
but with an Error method:
```go
var err interface{} = ExitError{1}
_, ok := err.(error) // ok is true
```

**Comment by kalman** [comment]
back to you.

@aboodman what do you mean by "factor out go/util/exit"?
* the code which clients are supposed to call to Exit in `d.Exit`. Being in the `d` package is pretty weird IMO, go/util/exit would make more sense.
* the code (which @zcstarr has just written) that deals specifically with exit code errors seems like it arguably does belong in the client test suite, because exit codes are a client concept. However, coupling this to a test suite does seem bad. Having in go/util/exit seems nice.

**Comment by aboodman** [comment]
This is looking great to me. I agree that the mock exit should ideally be in `go/util/error`.

I disagree that things that are 'client specific' should go in 'clients'. We're a peer-to-peer database. Everything we do is client-specific :). Less snarkily, it gets too confusing to try and figure out for every piece of code whether it will be used by "clients" or not. That is why I think all of this sort of thing should just go in `go/util` and let the linker figure it out.

Only thing IMO that should be in `samples/go` are actual sample binaries. I'd put everything else in `go/util`.

File: samples/go/poke/main_test.go (hunk 3 above)
Comment (zcstarr) [inline samples/go/poke/main_test.go]: Not sure I can here, if err is nil, I can't type assert it

File: samples/go/noms-merge/main_test.go (hunk 1 above)
Comment (zcstarr) [inline samples/go/noms-merge/main_test.go]: Simply didn't want to muck around with the existing test logic too much in this commit. I think @cmasone-attic  has an issue for me to refactor test a little bit to mesh better with noms-merge

File: samples/go/noms-merge/main_test.go (hunk 2 above)
Comment (zcstarr) [inline samples/go/noms-merge/main_test.go]: ditto

File: go/util/clienttest/client_test_suite.go (hunk 6 above)
Comment (zcstarr) [inline go/util/clienttest/client_test_suite.go]: good idea done.

**Comment by zcstarr** [comment]
@aboodman @kalman PTAL didn't add too much, just addressed the issues in the comments

File: go/util/clienttest/client_test_suite.go [hunk 7]
```diff
@@ -27,6 +27,11 @@ type ExitError struct {
 	Code int
 }
 
+func (e ExitError) Error() string {
+
```
Comment (aboodman) [inline go/util/clienttest/client_test_suite.go]: Remove extraneous line.

File: go/util/clienttest/client_test_suite.go [hunk 8]
```diff
@@ -48,6 +53,7 @@ func (suite *ClientTestSuite) TearDownSuite() {
 	defer d.Chk.NoError(os.RemoveAll(suite.TempDir))
 }
 
+//MustRun is a wrapper around Run that will panic on Exit or Panic
```
Comment (aboodman) [inline go/util/clienttest/client_test_suite.go]: Add a space after `//`

Comment (aboodman) [inline go/util/clienttest/client_test_suite.go]: Add a period at end of sentence.

File: go/util/clienttest/client_test_suite.go [hunk 9]
```diff
@@ -56,6 +62,8 @@ func (suite *ClientTestSuite) MustRun(m func(), args []string) (stdout string, s
 	return
 }
 
+//Run will execute a function passing args as commandline args, and recovers the function
```
Comment (aboodman) [inline go/util/clienttest/client_test_suite.go]: Spaces after '//'

File: go/util/clienttest/client_test_suite.go [hunk 10]
```diff
@@ -43,7 +53,18 @@ func (suite *ClientTestSuite) TearDownSuite() {
...
+	var err interface{}
+	if stdout, stderr, err = suite.Run(m, args); err != nil {
+		panic(err)
+	}
+	return
+}
+
+//Run will execute a function passing args as commandline args, and recovers the function
+//on Panic or os.Exit()
+func (suite *ClientTestSuite) Run(m func(), args []string) (stdout string, stderr string, mainErr interface{}) {
```
Comment (aboodman) [inline go/util/clienttest/client_test_suite.go]: I believe you can change `mainErr` to `error` type now.

**Comment by aboodman** [comment]
BTY, please try to make the return type `error`,.

**Comment by aboodman** [comment]
lgtm

**Comment by aboodman** [comment]
thanks!

**Pull request merged**